### PR TITLE
Share common code in 'use collection expression'.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -6,68 +6,43 @@
...
-                .Add(s_descriptor, CodeStyleOptions2.PreferCollectionExpression)
-                .Add(s_unnecessaryCodeDescriptor, CodeStyleOptions2.PreferCollectionExpression))
+        : base(IDEDiagnosticIds.UseCollectionExpressionForArrayDiagnosticId,
+               EnforceOnBuildValues.UseCollectionExpressionForArray)
     {
     }
 
-    protected override void InitializeWorker(AnalysisContext context)
-        => context.RegisterCompilationStartAction(context =>
+    protected override void InitializeWorker(CompilationStartAnalysisContext context)
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs:26]: view with whitespace off.

File: src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs [hunk 2]
```diff
@@ -6,68 +6,43 @@
...
-        new LocalizableResourceString(nameof(AnalyzersResources.Simplify_collection_initialization), AnalyzersResources.ResourceManager, typeof(AnalyzersResources)),
-        new LocalizableResourceString(nameof(AnalyzersResources.Collection_initialization_can_be_simplified), AnalyzersResources.ResourceManager, typeof(AnalyzersResources)),
-        isUnnecessary: true);
-
     public CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer()
-        : base(ImmutableDictionary<DiagnosticDescriptor, IOption2>.Empty
-                .Add(s_descriptor, CodeStyleOptions2.PreferCollectionExpression)
-                .Add(s_unnecessaryCodeDescriptor, CodeStyleOptions2.PreferCollectionExpression))
+        : base(IDEDiagnosticIds.UseCollectionExpressionForArrayDiagnosticId,
+               EnforceOnBuildValues.UseCollectionExpressionForArray)
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs:22]: each subclass needs to pass in these values to make sure the analyzer applies to it.

File: src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs [hunk 3]
```diff
@@ -6,68 +6,43 @@
...
+    {
+        // We wrap the SyntaxNodeAction within a CodeBlockStartAction, which allows us to
+        // get callbacks for object creation expression nodes, but analyze nodes across the entire code block
+        // and eventually report fading diagnostics with location outside this node.
+        // Without the containing CodeBlockStartAction, our reported diagnostic would be classified
+        // as a non-local diagnostic and would not participate in lightbulb for computing code fixes.
+        context.RegisterCodeBlockStartAction<SyntaxKind>(context =>
         {
-            if (!context.Compilation.LanguageVersion().SupportsCollectionExpressions())
-                return;
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs:51]: base type does this check as it's common to all cases.

File: src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs [hunk 4]
```diff
@@ -6,68 +6,30 @@
...
-                    context => AnalyzeArrayCreationExpression(context),
-                    SyntaxKind.ArrayCreationExpression);
-            });
-        });
-
-    private static void AnalyzeArrayCreationExpression(SyntaxNodeAnalysisContext context)
+    protected override void InitializeWorker(CodeBlockStartAnalysisContext<SyntaxKind> context)
+    {
+        context.RegisterSyntaxNodeAction(AnalyzeArrayInitializerExpression, SyntaxKind.ArrayInitializerExpression);
+        context.RegisterSyntaxNodeAction(AnalyzeArrayCreationExpression, SyntaxKind.ArrayCreationExpression);
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs:29]: subtypes only ned to do this registration.  all the outside parts are handled by the base type.

File: src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs [hunk 5]
```diff
@@ -6,68 +6,30 @@
...
+    }
+
+    private void AnalyzeArrayCreationExpression(SyntaxNodeAnalysisContext context)
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs:32]: the descriptors are now instance state (As they are instance specific).  so these methods became non-static.

**Review by akhera99 — APPROVED** [review]

**Pull request merged**